<a href="https://colab.research.google.com/github/wyacheslav-20/AIStrel/blob/main/nb/Qwen3_(4B)-GRPO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Unsloth now supports Text-to-Speech (TTS) models. Read our [guide here](https://docs.unsloth.ai/basics/text-to-speech-tts-fine-tuning).

Read our **[Gemma 3N Guide](https://docs.unsloth.ai/basics/gemma-3n-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [22]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    !pip install --no-deps unsloth vllm==0.8.5.post1

In [23]:
#@title Colab Extra Install { display-mode: "form" }
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    !pip install --no-deps unsloth vllm==0.8.5.post1
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    # Skip restarting message in Colab
    import sys, re, requests; modules = list(sys.modules.keys())
    for x in modules: sys.modules.pop(x) if "PIL" in x or "google" in x else None
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets>=3.4.1,<4.0.0" "huggingface_hub>=0.34.0" hf_transfer

    # vLLM requirements - vLLM breaks Colab due to reinstalling numpy
    f = requests.get("https://raw.githubusercontent.com/vllm-project/vllm/refs/heads/main/requirements/common.txt").content
    with open("vllm_requirements.txt", "wb") as file:
        file.write(re.sub(rb"(transformers|numpy|xformers)[^\n]{1,}\n", b"", f))
    !pip install -r vllm_requirements.txt

### Unsloth

Goal: To convert `Qwen3-4B-Base` into a reasoning model via GRPO by using OpenR1's Math dataset.

We first pre fine-tune the model to make GRPO skip trying to match formatting - this speeds GRPO up.

In [24]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Can increase for longer reasoning traces
lora_rank = 32 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/Qwen3-0.6B-Base",
    max_seq_length = max_seq_length,
    load_in_4bit = False, # False for LoRA 16bit
    fast_inference = True, # Enable vLLM fast inference
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.7, # Reduce if out of memory
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = [
        "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj",
    ],
    lora_alpha = lora_rank*2, # *2 speeds up training
    use_gradient_checkpointing = "unsloth", # Reduces memory usage
    random_state = 3407,
)

Unsloth: Patching vLLM v1 graph capture
Unsloth: Patching vLLM v0 graph capture
==((====))==  Unsloth 2025.8.1: Fast Qwen3 patching. Transformers: 4.54.0. vLLM: 0.8.5.post1.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: vLLM loading unsloth/Qwen3-0.6B-Base with actual GPU utilization = 52.62%
Unsloth: Your GPU has CUDA compute capability 7.5 with VRAM = 14.74 GB.
Unsloth: Using conservativeness = 1.0. Chunked prefill tokens = 2048. Num Sequences = 192.
Unsloth: vLLM's KV Cache can use up to 6.78 GB. Also swap space = 0 GB.
WARNING 08-05 16:07:43 [config.py:2972] Casting torch.bfloat16 to torch.float16.
INFO 08-05 16:07:43 [config.py:717] This model

Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


INFO 08-05 16:07:51 [loader.py:458] Loading weights took 2.16 seconds
INFO 08-05 16:07:52 [model_runner.py:1140] Model loading took 1.1551 GiB and 2.617706 seconds
INFO 08-05 16:07:55 [worker.py:287] Memory profiling takes 1.55 seconds
INFO 08-05 16:07:55 [worker.py:287] the current vLLM instance can use total_gpu_memory (14.74GiB) x gpu_memory_utilization (0.53) = 7.76GiB
INFO 08-05 16:07:55 [worker.py:287] model weights take 1.16GiB; non_torch_memory takes 0.00GiB; PyTorch activation peak memory takes 1.04GiB; the rest of the memory reserved for KV Cache is 5.56GiB.
INFO 08-05 16:07:56 [executor_base.py:112] # cuda blocks: 3254, # CPU blocks: 0
INFO 08-05 16:07:56 [executor_base.py:117] Maximum concurrency for 2048 tokens per request: 25.42x
INFO 08-05 16:07:56 [vllm_utils.py:669] Unsloth: Running patched vLLM v0 `capture_model`.
INFO 08-05 16:07:56 [vllm_utils.py:669] Unsloth: Running patched vLLM v0 `capture_model`.
INFO 08-05 16:07:56 [vllm_utils.py:669] Unsloth: Running patched v

Capturing CUDA graph shapes:   0%|          | 0/27 [00:00<?, ?it/s]

INFO 08-05 16:08:02 [model_runner.py:1592] Graph capturing finished in 7 secs, took 0.04 GiB
INFO 08-05 16:08:02 [vllm_utils.py:676] Unsloth: Patched vLLM v0 graph capture finished in 7 secs.
INFO 08-05 16:08:02 [vllm_utils.py:676] Unsloth: Patched vLLM v0 graph capture finished in 7 secs.
INFO 08-05 16:08:02 [vllm_utils.py:676] Unsloth: Patched vLLM v0 graph capture finished in 7 secs.
INFO 08-05 16:08:04 [llm_engine.py:437] init engine (profile, create kv cache, warmup model) took 11.10 seconds
Unsloth: Just some info: will skip parsing ['pre_feedforward_layernorm', 'post_feedforward_layernorm']
Unsloth: Just some info: will skip parsing ['pre_feedforward_layernorm', 'post_feedforward_layernorm']


### GRPO chat template
Since we're using a base model, we should set a chat template. You can make your own chat template as well!
1. DeepSeek uses `<think>` and `</think>`, but this is **not** necessary - you can customize it however you like!
2. A `system_prompt` is recommended to at least guide the model's responses.

In [25]:
reasoning_start = "<start_working_out>" # Acts as <think>
reasoning_end   = "<end_working_out>"   # Acts as </think>
solution_start  = "<SOLUTION>"
solution_end    = "</SOLUTION>"

system_prompt = \
f"""You are given a problem.
Think about the problem and provide your working out.
Place it between {reasoning_start} and {reasoning_end}.
Then, provide your solution between {solution_start}{solution_end}"""
system_prompt

'You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION>'

We create a simple chat template below. Notice `add_generation_prompt` includes prepending `<start_working_out>` to guide the model to start its reasoning process.

In [26]:
chat_template = \
    "{% if messages[0]['role'] == 'system' %}"\
        "{{ messages[0]['content'] + eos_token }}"\
        "{% set loop_messages = messages[1:] %}"\
    "{% else %}"\
        "{{ '{system_prompt}' + eos_token }}"\
        "{% set loop_messages = messages %}"\
    "{% endif %}"\
    "{% for message in loop_messages %}"\
        "{% if message['role'] == 'user' %}"\
            "{{ message['content'] }}"\
        "{% elif message['role'] == 'assistant' %}"\
            "{{ message['content'] + eos_token }}"\
        "{% endif %}"\
    "{% endfor %}"\
    "{% if add_generation_prompt %}{{ '{reasoning_start}' }}"\
    "{% endif %}"

# Replace with out specific template:
chat_template = chat_template\
    .replace("'{system_prompt}'",   f"'{system_prompt}'")\
    .replace("'{reasoning_start}'", f"'{reasoning_start}'")
tokenizer.chat_template = chat_template

Let's see how our chat template behaves on an example:

In [27]:
tokenizer.apply_chat_template([
    {"role" : "user", "content" : "What is 1+1?"},
    {"role" : "assistant", "content" : f"{reasoning_start}I think it's 2.{reasoning_end}{solution_start}2{solution_end}"},
    {"role" : "user", "content" : "What is 2+2?"},
], tokenize = False, add_generation_prompt = True)

"You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION><|endoftext|>What is 1+1?<start_working_out>I think it's 2.<end_working_out><SOLUTION>2</SOLUTION><|endoftext|>What is 2+2?<start_working_out>"

### Pre fine-tuning for formatting
We now use a subset of NVIDIA's [Open Math Reasoning dataset](https://huggingface.co/datasets/nvidia/OpenMathReasoning) which was filtered to only include high quality DeepSeek R1 traces.

We'll only filter ~59 or so examples to first "prime" / pre fine-tune the model to understand our custom GRPO formatting.

In [28]:
from datasets import load_dataset
import pandas as pd
import numpy as np

dataset = load_dataset("unsloth/OpenMathReasoning-mini", split = "cot")
dataset = dataset.to_pandas()[
    ["expected_answer", "problem", "generated_solution"]
]

# Try converting to number - if not, replace with NaN
is_number = pd.to_numeric(pd.Series(dataset["expected_answer"]), errors = "coerce").notnull()
# Select only numbers
dataset = dataset.iloc[np.where(is_number)[0]]

dataset

README.md:   0%|          | 0.00/603 [00:00<?, ?B/s]

data/cot-00000-of-00001.parquet:   0%|          | 0.00/106M [00:00<?, ?B/s]

Generating cot split:   0%|          | 0/19252 [00:00<?, ? examples/s]

expected_answer                                            problem  \
0                  14  Given $\sqrt{x^2+165}-\sqrt{x^2-52}=7$ and $x$...   
6                  -2  Find the value of the parameter $a$ for which ...   
9                  18  What is the sum of all real numbers $x$ for wh...   
13                  2  Evaluate the sum \(\sum_{n=1}^\infty \frac{\ph...   
17                 30  What is the largest positive integer that divi...   
...               ...                                                ...   
19243             244  Let \( p \), \( q \), and \( r \) be the disti...   
19245               1  A bug is on the $0$ of a number line. At any p...   
19247               4  A bus left point X for point Y. Two hours late...   
19248              18  Each interior angle of a regular n-gon measure...   
19250          0.8960  Find the probability that the second blue resu...   

                                      generated_solution  
0      <think>\nOkay, let's see. I need to solve the ...  
6      <think>\nOkay, so I need to find the value of ...  
9      <think>\nOkay, so I need to solve the equation...  
13     <think>\nOkay, so I need to evaluate the infin...  
17     <think>\nAlright, so I need to find the larges...  
...                                                  ...  
19243  <think>\nOkay, so I need to find the value of ...  
19245  <think>\nOkay, so I have this problem where a ...  
19247  <think>\nOkay, let's tackle this problem step ...  
19248  <think>\nOkay, let's see. I need to find the n...  
19250  <think>\nOkay, so I need to find the probabili...  

[7507 rows x 3 columns]

We have to format the dataset to follow our GRPO style formatting:

In [29]:
def format_dataset(x):
    expected_answer = x["expected_answer"]
    problem = x["problem"]

    # Remove generated <think> and </think>
    thoughts = x["generated_solution"]
    thoughts = thoughts.replace("<think>", "").replace("</think>", "")

    # Strip newlines on left and right
    thoughts = thoughts.strip()
    # Add our custom formatting
    final_prompt = \
        reasoning_start + thoughts + reasoning_end + \
        solution_start + expected_answer + solution_end
    return [
        {"role" : "system",    "content" : system_prompt},
        {"role" : "user",      "content" : problem},
        {"role" : "assistant", "content" : final_prompt},
    ]

dataset["Messages"] = dataset.apply(format_dataset, axis = 1)

Check to see if it worked:

In [30]:
tokenizer.apply_chat_template(dataset["Messages"][0], tokenize = False)

"You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION><|endoftext|>Given $\\sqrt{x^2+165}-\\sqrt{x^2-52}=7$ and $x$ is positive, find all possible values of $x$.<start_working_out>Okay, let's see. I need to solve the equation √(x² + 165) - √(x² - 52) = 7, and find all positive values of x. Hmm, radicals can be tricky, but maybe if I can eliminate the square roots by squaring both sides. Let me try that.\n\nFirst, let me write down the equation again to make sure I have it right:\n\n√(x² + 165) - √(x² - 52) = 7.\n\nOkay, so the idea is to isolate one of the radicals and then square both sides. Let me try moving the second radical to the other side:\n\n√(x² + 165) = 7 + √(x² - 52).\n\nNow, if I square both sides, maybe I can get rid of the square roots. Let's do that:\n\n(√(x² + 165))² = (7 + √(x² - 52))².\n\nSimplifying the left side:\n\nx² + 

Let's truncate the pre fine-tuning dataset to `max_seq_length/2` since we don't want too long reasoning traces.

Note this might take 2 minutes!

In [31]:
dataset["N"] = dataset["Messages"].apply(lambda x: len(tokenizer.apply_chat_template(x)))

dataset = dataset.loc[dataset["N"] <= max_seq_length/2].copy()
dataset.shape

(59, 5)

We then tokenize the messages and convert it to a Hugging Face compatible dataset format:

In [32]:
from datasets import Dataset

dataset["text"] = tokenizer.apply_chat_template(dataset["Messages"].values.tolist(), tokenize = False)
dataset = Dataset.from_pandas(dataset)
dataset

Dataset({
    features: ['expected_answer', 'problem', 'generated_solution', 'Messages', 'N', 'text', '__index_level_0__'],
    num_rows: 59
})

Let's now pre fine-tune the model so it follows our custom GRPO formatting!

In [33]:
from trl import SFTTrainer, SFTConfig
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    args = SFTConfig(
        dataset_text_field = "text",
        per_device_train_batch_size = 1,
        gradient_accumulation_steps = 1, # Use GA to mimic batch size!
        warmup_steps = 5,
        num_train_epochs = 2, # Set this for 1 full training run.
        learning_rate = 2e-4, # Reduce to 2e-5 for long training runs
        logging_steps = 5,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        report_to = "none", # Use this for WandB etc
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/59 [00:00<?, ? examples/s]

In [34]:
trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 59 | Num Epochs = 2 | Total steps = 118
O^O/ \_/ \    Batch size per device = 1 | Gradient accumulation steps = 1
\        /    Data Parallel GPUs = 1 | Total batch size (1 x 1 x 1) = 1
 "-____-"     Trainable parameters = 20,185,088 of 616,235,008 (3.28% trained)


TrainOutput(global_step=118, training_loss=0.4827902852478674, metrics={'train_runtime': 54.6136, 'train_samples_per_second': 2.161, 'train_steps_per_second': 2.161, 'total_flos': 295622848806912.0, 'train_loss': 0.4827902852478674})

Let's check if the model has learnt to follow the custom format:

In [35]:
text = tokenizer.apply_chat_template(
    dataset[0]["Messages"][:2],
    tokenize = False,
    add_generation_prompt = True, # Must add for generation
)

from transformers import TextStreamer
_ = model.generate(
    **tokenizer(text, return_tensors = "pt").to("cuda"),
    temperature = 0,
    max_new_tokens = 1024,
    streamer = TextStreamer(tokenizer, skip_prompt = False),
)

You are given a problem.
Think about the problem and provide your working out.
Place it between <start_working_out> and <end_working_out>.
Then, provide your solution between <SOLUTION></SOLUTION><|endoftext|>Jenifer has 82 cents in pennies and nickels. Her younger brother mistook all her nickels for dimes and counted the total as $1.47. How many pennies does Jenifer have?<start_working_out>Okay, let's see. So Jenifer has 82 cents in pennies and nickels. Her brother thought all the nickels were dimes and counted the total as $1.47. I need to find out how many pennies she has. Hmm, let's break this down.

First, I need to remember the value of each coin. A penny is 1 cent, a nickel is 5 cents, a dime is 10 cents, and a quarter is 25 cents. But in this problem, the brother thought the nickels were dimes. So maybe he was counting the nickels as dimes instead of pennies. Let me check that.

If the nickels were dimes, then each nickel would be worth 10 cents, and each dime would be 20 cents

Yes it did follow the formatting! Great! Let's remove some items before the GRPO step

In [36]:
del dataset
torch.cuda.empty_cache()
import gc
gc.collect()

139

### Data Prep
<a name="Data"></a>

We're using Hugging Face's [Open R1 Math dataset](https://huggingface.co/datasets/open-r1/DAPO-Math-17k-Processed). You can also utilize OpenAI's famous [GSM8K dataset](https://huggingface.co/datasets/openai/gsm8k)

In [37]:
from datasets import load_dataset
dataset = load_dataset("swa19511951/AlexCtrel", split = "train")
dataset

Dataset({
    features: ['text'],
    num_rows: 9010
})

Let's look at the first row:

In [ ]:
dataset[0]["prompt"]

'In triangle $ABC$, $\\sin \\angle A = \\frac{4}{5}$ and $\\angle A < 90^\\circ$. Let $D$ be a point outside triangle $ABC$ such that $\\angle BAD = \\angle DAC$ and $\\angle BDC = 90^\\circ$. Suppose that $AD = 1$ and that $\\frac{BD}{CD} = \\frac{3}{2}$. If $AB + AC$ can be expressed in the form $\\frac{a\\sqrt{b}}{c}$ where $a, b, c$ are pairwise relatively prime integers, find $a + b + c$.'

In [40]:
dataset[0]["text"]

'<|im_start|>system\nYou are a helpful assistant.\n<|im_end|>\n<|im_start|>user\nÐ\x9aакой оÑ\x81новной вопÑ\x80оÑ\x81 обÑ\x81Ñ\x83ждаеÑ\x82Ñ\x81Ñ\x8f в Ñ\x82екÑ\x81Ñ\x82е?\n<|im_end|>\n<|im_start|>assistant<|thinking|>\nÐ\x9fеÑ\x80вое пÑ\x80едложение пÑ\x80едÑ\x81Ñ\x82авлÑ\x8fеÑ\x82 Ñ\x81обой иÑ\x81Ñ\x85однÑ\x8bй вопÑ\x80оÑ\x81 коÑ\x82оÑ\x80Ñ\x8bй авÑ\x82оÑ\x80 Ñ\x85оÑ\x87еÑ\x82 обÑ\x81Ñ\x83диÑ\x82Ñ\x8c в Ñ\x82екÑ\x81Ñ\x82е.\n<|im_end|>\n<|im_start|>user\nЯ Ñ\x85оÑ\x82ел бÑ\x8b оÑ\x82веÑ\x82иÑ\x82Ñ\x8c на вопÑ\x80оÑ\x81 наÑ\x81колÑ\x8cко белоÑ\x80Ñ\x83Ñ\x81Ñ\x81кие Ñ\x80ежиÑ\x81Ñ\x81еÑ\x80Ñ\x8b опÑ\x80еделÑ\x8fÑ\x8eÑ\x82 лиÑ\x86о белоÑ\x80Ñ\x83Ñ\x81Ñ\x81кого Ñ\x82еаÑ\x82Ñ\x80а\n<|im_end|>'

In GSM8K, ee notice all answers like about have a ####, so we extract it. But for the Open R1 dataset, we can skip the below.

In [42]:
def extract_hash_answer(text):
    # if "####" not in text: return None
    # return text.split("####")[1].strip()
    return text
extract_hash_answer(dataset[0]["text"])

'<|im_start|>system\nYou are a helpful assistant.\n<|im_end|>\n<|im_start|>user\nÐ\x9aакой оÑ\x81новной вопÑ\x80оÑ\x81 обÑ\x81Ñ\x83ждаеÑ\x82Ñ\x81Ñ\x8f в Ñ\x82екÑ\x81Ñ\x82е?\n<|im_end|>\n<|im_start|>assistant<|thinking|>\nÐ\x9fеÑ\x80вое пÑ\x80едложение пÑ\x80едÑ\x81Ñ\x82авлÑ\x8fеÑ\x82 Ñ\x81обой иÑ\x81Ñ\x85однÑ\x8bй вопÑ\x80оÑ\x81 коÑ\x82оÑ\x80Ñ\x8bй авÑ\x82оÑ\x80 Ñ\x85оÑ\x87еÑ\x82 обÑ\x81Ñ\x83диÑ\x82Ñ\x8c в Ñ\x82екÑ\x81Ñ\x82е.\n<|im_end|>\n<|im_start|>user\nЯ Ñ\x85оÑ\x82ел бÑ\x8b оÑ\x82веÑ\x82иÑ\x82Ñ\x8c на вопÑ\x80оÑ\x81 наÑ\x81колÑ\x8cко белоÑ\x80Ñ\x83Ñ\x81Ñ\x81кие Ñ\x80ежиÑ\x81Ñ\x81еÑ\x80Ñ\x8b опÑ\x80еделÑ\x8fÑ\x8eÑ\x82 лиÑ\x86о белоÑ\x80Ñ\x83Ñ\x81Ñ\x81кого Ñ\x82еаÑ\x82Ñ\x80а\n<|im_end|>'

Let's map the dataset! and see the first row:

In [44]:
dataset = dataset.map(lambda x: {
    "prompt" : [
        {"role": "system", "content": system_prompt},
        {"role": "user",   "content": x["text"]},
    ],
    "answer": extract_hash_answer(x["text"]),
})
dataset[0]

Map:   0%|          | 0/9010 [00:00<?, ? examples/s]

{'text': '<|im_start|>system\nYou are a helpful assistant.\n<|im_end|>\n<|im_start|>user\nÐ\x9aакой оÑ\x81новной вопÑ\x80оÑ\x81 обÑ\x81Ñ\x83ждаеÑ\x82Ñ\x81Ñ\x8f в Ñ\x82екÑ\x81Ñ\x82е?\n<|im_end|>\n<|im_start|>assistant<|thinking|>\nÐ\x9fеÑ\x80вое пÑ\x80едложение пÑ\x80едÑ\x81Ñ\x82авлÑ\x8fеÑ\x82 Ñ\x81обой иÑ\x81Ñ\x85однÑ\x8bй вопÑ\x80оÑ\x81 коÑ\x82оÑ\x80Ñ\x8bй авÑ\x82оÑ\x80 Ñ\x85оÑ\x87еÑ\x82 обÑ\x81Ñ\x83диÑ\x82Ñ\x8c в Ñ\x82екÑ\x81Ñ\x82е.\n<|im_end|>\n<|im_start|>user\nЯ Ñ\x85оÑ\x82ел бÑ\x8b оÑ\x82веÑ\x82иÑ\x82Ñ\x8c на вопÑ\x80оÑ\x81 наÑ\x81колÑ\x8cко белоÑ\x80Ñ\x83Ñ\x81Ñ\x81кие Ñ\x80ежиÑ\x81Ñ\x81еÑ\x80Ñ\x8b опÑ\x80еделÑ\x8fÑ\x8eÑ\x82 лиÑ\x86о белоÑ\x80Ñ\x83Ñ\x81Ñ\x81кого Ñ\x82еаÑ\x82Ñ\x80а\n<|im_end|>',
 'prompt': [{'content': 'You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_o

We create a regex format to match the reasoning sections and answers:

In [45]:
import re

# Add optional EOS token matching
solution_end_regex = r"</SOLUTION>[\s]{0,}" + \
    "(?:" + re.escape(tokenizer.eos_token) + ")?"

match_format = re.compile(
    rf"{reasoning_end}.*?"\
    rf"{solution_start}(.+?){solution_end_regex}"\
    rf"[\s]{{0,}}$",
    flags = re.MULTILINE | re.DOTALL
)
match_format

re.compile(r'<end_working_out>.*?<SOLUTION>(.+?)</SOLUTION>[\s]{0,}(?:<\|endoftext\|>)?[\s]{0,}$',
re.MULTILINE|re.DOTALL|re.UNICODE)

We verify it works:

In [46]:
match_format.findall(
    "Let me think!<end_working_out>"\
    f"<SOLUTION>\n2\n</SOLUTION>",
)

['\n2\n']

In [47]:
match_format.findall(
    "<start_working_out>Let me think!<end_working_out>"\
    f"<SOLUTION>  2  </SOLUTION>\n\n",
)

['  2  ']

We now want to create a reward function to match the format exactly - we reward it with 3 points if it succeeds:

In [48]:
def match_format_exactly(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Match if format is seen exactly!
        if match_format.search(response) is not None: score += 3.0
        scores.append(score)
    return scores

If it fails, we want to reward the model if it at least follows the format partially, by counting each symbol:

In [49]:
def match_format_approximately(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Count how many keywords are seen - we penalize if too many!
        # If we see 1, then plus some points!

        # No need to reward <start_working_out> since we always prepend it!
        # score += 0.5 if response.count(reasoning_start) == 1 else -1.0
        score += 0.5 if response.count(reasoning_end)   == 1 else -1.0
        score += 0.5 if response.count(solution_start)  == 1 else -1.0
        score += 0.5 if response.count(solution_end)    == 1 else -1.0
        scores.append(score)
    return scores

Finally, we want to extract the generated answer, and reward or penalize it! We also reward it based on how close the answer is to the true one via ratios:

In [50]:
def check_answer(prompts, completions, answer, **kwargs):
    question = prompts[0][-1]["content"]
    responses = [completion[0]["content"] for completion in completions]

    extracted_responses = [
        guess.group(1)
        if (guess := match_format.search(r)) is not None else None \
        for r in responses
    ]

    scores = []
    for guess, true_answer in zip(extracted_responses, answer):
        score = 0
        if guess is None:
            scores.append(-2.0)
            continue
        # Correct answer gets 5 points!
        if guess == true_answer:
            score += 5.0
        # Match if spaces are seen, but less reward
        elif guess.strip() == true_answer.strip():
            score += 3.5
        else:
            # We also reward it if the answer is close via ratios!
            # Ie if the answer is within some range, reward it!
            try:
                ratio = float(guess) / float(true_answer)
                if   ratio >= 0.9 and ratio <= 1.1: score += 2.0
                elif ratio >= 0.8 and ratio <= 1.2: score += 1.5
                else: score -= 2.5 # Penalize wrong answers
            except:
                score -= 4.5 # Penalize
        scores.append(score)
    return scores

Also sometimes it might not be 1 number as the answer, but like a sentence for example "The solution is $20" -> we extract 20.

We also remove possible commas for example as in 123,456

In [51]:
match_numbers = re.compile(
    solution_start + r".*?[\s]{0,}([-]?[\d\.\,]{1,})",
    flags = re.MULTILINE | re.DOTALL
)
print(match_numbers.findall("<SOLUTION>  0.34  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>  123,456  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>  -0.234  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>17</SOLUTION>"))

['0.34']
['123,456']
['-0.234']
['17']


We now prepare our main function which will print out the generated responses and the true answer, along with another reward function which converts text to float via `float` and sees if it's the same.

In [52]:
global PRINTED_TIMES
PRINTED_TIMES = 0
global PRINT_EVERY_STEPS
PRINT_EVERY_STEPS = 5

def check_numbers(prompts, completions, answer, **kwargs):
    question = prompts[0][-1]["content"]
    responses = [completion[0]["content"] for completion in completions]

    extracted_responses = [
        guess.group(1)
        if (guess := match_numbers.search(r)) is not None else None \
        for r in responses
    ]

    scores = []
    # Print only every few steps
    global PRINTED_TIMES
    global PRINT_EVERY_STEPS
    if PRINTED_TIMES % PRINT_EVERY_STEPS == 0:
        print(
            '*'*20 + f"Question:\n{question}", f"\nAnswer:\n{answer[0]}", f"\nResponse:\n{responses[0]}", f"\nExtracted:\n{extracted_responses[0]}"
        )
    PRINTED_TIMES += 1

    for guess, true_answer in zip(extracted_responses, answer):
        if guess is None:
            scores.append(-2.5)
            continue
        # Convert to numbers
        try:
            true_answer = float(true_answer.strip())
            # Remove commas like in 123,456
            guess       = float(guess.strip().replace(",", ""))
            scores.append(3.5 if guess == true_answer else -1.5)
        except:
            scores.append(0)
            continue
    return scores

Get the top 90% prompt length so we don't accidentally truncate them!

Ie we'll remove the top 10% long prompts.

In [53]:
tokenized = dataset.map(
    lambda x: {"tokens" : tokenizer.apply_chat_template(x["prompt"], add_generation_prompt = True, tokenize = True)},
    batched = True,
)
print(tokenizer.decode(tokenized[0]["tokens"]))
tokenized = tokenized.map(lambda x: {"L" : len(x["tokens"])})

import numpy as np
maximum_length = int(np.quantile(tokenized["L"], 0.9))
print("Max Length = ", maximum_length)

# Filter only samples smaller than 90% max length
dataset = dataset.select(np.where(np.array(tokenized["L"]) <= maximum_length)[0])
del tokenized

Map:   0%|          | 0/9010 [00:00<?, ? examples/s]

You are given a problem.
Think about the problem and provide your working out.
Place it between <start_working_out> and <end_working_out>.
Then, provide your solution between <SOLUTION></SOLUTION><|endoftext|><|im_start|>system
You are a helpful assistant.
<|im_end|>
<|im_start|>user
Какой основной вопрос обсуждается в тексте?
<|im_end|>
<|im_start|>assistant<|thinking|>
Первое предложение представляет собой исходный вопрос который автор хочет обсудить в тексте.
<|im_end|>
<|im_start|>user
Я хотел бы ответить на вопрос насколько белорусские режиссеры определяют лицо белорусского театра
<|im_end|><start_working_out>


Map:   0%|          | 0/9010 [00:00<?, ? examples/s]

Max Length =  777


<a name="Train"></a>
### Train the model

Now set up GRPO Trainer and all configurations!

In [54]:
max_prompt_length = maximum_length + 1 # + 1 just in case!
max_completion_length = max_seq_length - max_prompt_length

from vllm import SamplingParams
vllm_sampling_params = SamplingParams(
    min_p = 0.1,
    top_p = 1.0,
    top_k = -1,
    seed = 3407,
    stop = [tokenizer.eos_token],
    include_stop_str_in_output = True,
)

from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    vllm_sampling_params = vllm_sampling_params,
    temperature = 1.0,
    learning_rate = 5e-6,
    weight_decay = 0.01,
    warmup_ratio = 0.1,
    lr_scheduler_type = "linear",
    optim = "adamw_8bit",
    logging_steps = 1,
    per_device_train_batch_size = 1,
    gradient_accumulation_steps = 1, # Increase to 4 for smoother training
    num_generations = 4, # Decrease if out of memory
    max_prompt_length = max_prompt_length,
    max_completion_length = max_completion_length,
    # num_train_epochs = 1, # Set to 1 for a full training run
    max_steps = 100,
    save_steps = 100,
    report_to = "none", # Can use Weights & Biases
    output_dir = "outputs",

    # For optional training + evaluation
    # fp16_full_eval = True,
    # per_device_eval_batch_size = 4,
    # eval_accumulation_steps = 1,
    # eval_strategy = "steps",
    # eval_steps = 1,
)

Unsloth: We now expect `per_device_train_batch_size` to be a multiple of `num_generations`.
We will change the batch size of 1 to the `num_generations` of 4


And let's run the trainer! If you scroll up, you'll see a table of rewards. The goal is to see the `reward` column increase!

You might have to wait 150 to 200 steps for any action. You'll probably get 0 reward for the first 100 steps. Please be patient!

| Step | Training Loss | reward    | reward_std | completion_length | kl       |
|------|---------------|-----------|------------|-------------------|----------|
| 1    | 0.000000      | 0.125000  | 0.000000   | 200.000000        | 0.000000 |
| 2    | 0.000000      | 0.072375  | 0.248112   | 200.000000        | 0.000000 |
| 3    | 0.000000      | -0.079000 | 0.163776   | 182.500000        | 0.000005 |


In [55]:
# For optional training + evaluation
# new_dataset = dataset.train_test_split(test_size = 0.01)

trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        match_format_exactly,
        match_format_approximately,
        check_answer,
        check_numbers,
    ],
    args = training_args,
    train_dataset = dataset,

    # For optional training + evaluation
    # train_dataset = new_dataset["train"],
    # eval_dataset = new_dataset["test"],
)
trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 8,111 | Num Epochs = 1 | Total steps = 100
O^O/ \_/ \    Batch size per device = 4 | Gradient accumulation steps = 1
\        /    Data Parallel GPUs = 1 | Total batch size (4 x 1 x 1) = 4
 "-____-"     Trainable parameters = 20,185,088 of 616,235,008 (3.28% trained)


********************Question:
<|im_start|>system
You are a helpful assistant.
<|im_end|>
<|im_start|>user
Что Алексей Стрельников считает что интернет должен изменить?
<|im_end|>
<|im_start|>assistant<|thinking|>
интернет должен изменить многое.
<|im_end|>
<|im_start|>user
nan
<|im_end|> 
Answer:
<|im_start|>system
You are a helpful assistant.
<|im_end|>
<|im_start|>user
Что Алексей Стрельников считает что интернет должен изменить?
<|im_end|>
<|im_start|>assistant<|thinking|>
интернет должен изменить многое.
<|im_end|>
<|im_start|>user
nan
<|im_end|> 
Response:
Okay, let's see. The problem says: "ическогоsystem" You mentioned "благоцировогомноголетикуюсвязь", which I think is a combination of four words: "благовестничать", "кожухоретка", "людоежица" (which is "старица"; a female older person in Russian / Eastern European/Russi

********************Question:
<|im_start|>system
You are a helpful assistant.
<|im_end|>
<|im_start|>user
"Какой эффект может вызвать «раздражающе медленное» появление персонажей?"
<|im_end|>
<|im_start|>assistant<|thinking|>
"Создает напряжение или ожидание что может перерождаться в художественную интригу."
<|im_end|>
<|im_start|>user
nan
<|im_end|> 
Answer:
<|im_start|>system
You are a helpful assistant.
<|im_end|>
<|im_start|>user
"Какой эффект может вызвать «раздражающе медленное» появление персонажей?"
<|im_end|>
<|im_start|>assistant<|thinking|>
"Создает напряжение или ожидание что может перерождаться в художественную интригу."
<|im_end|>
<|im_start|>user
nan
<|im_end|> 
Response:
Okay, let me think about this prob

TrainOutput(global_step=100, training_loss=0.00026953320935717783, metrics={'train_runtime': 1899.8078, 'train_samples_per_second': 0.211, 'train_steps_per_second': 0.053, 'total_flos': 0.0, 'train_loss': 0.00026953320935717783})

<a name="Inference"></a>
### Inference
Now let's try the model we just trained! First, let's first try the model without any GRPO trained:

In [56]:
text = "Кто такой Алексей Стрельников?"

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 1024,
)
output = model.fast_generate(
    [text],
    sampling_params = sampling_params,
    lora_request = None,
)[0].outputs[0].text

output

Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

' Как ему удалось создать иностранный шлагбаум? Какой аспект его жизни стал резонансным в культурной и политической традиции России? Уже после выхода этого романа есть у него какие-то предрекнутые события?\n\nАлексей Стрельников — это несравненно молодой автор в творчестве, который присоединился к российской литературе в основном девятнадцатого и произведет последнее, что может выдать новую чушь как автора. Несмотря на то что в романе описывается его жизнь и рода Алексея, произведение имеет более низкую структуру и тему, чем другие экспериментальные ромы Стрельника.\n\nМнение о Стрельнике в последнее время очень разный, несмотря на его авторские права, поэтому нам не известно его идеи самого автора. Хотя не имеется никаких сведений о его деятельности за рубежом, можно обнародовать, что автор увлечён моральными вымырями, что в некоторых статьях можно перенести на обобщаю юмор, традиционное идее «чего-то подчиняется по крайней мере одной основной причине, и чем больше подчиняется, тем си

And now with the LoRA we just trained with GRPO - we first save the LoRA first!

In [57]:
model.save_lora("grpo_saved_lora")

Verify LoRA is actually trained!

In [58]:
from safetensors import safe_open

tensors = {}
with safe_open("grpo_saved_lora/adapter_model.safetensors", framework = "pt") as f:
    # Verify both A and B are non zero
    for key in f.keys():
        tensor = f.get_tensor(key)
        n_zeros = (tensor == 0).sum() / tensor.numel()
        assert(n_zeros.item() != tensor.numel())

Now we load the LoRA and test:

In [60]:
messages = [
    {"role": "system", "content": system_prompt},
    {"role": "user",   "content": "Кто такой Алексей Стрельников?"},
]

text = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    tokenize = False,
)
from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 2048,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)[0].outputs[0].text

output

Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

'Okay, so the problem is asking, "Кто такой Алексей Стрельников?" Which translates to "Who is Andrei Strilevsky?" Hmm, let me think. Andrei Strilevsky was a Soviet diplomat and politician who served in several high-ranking positions, including as the Minister of Internal Affairs and later as a diplomat in various regions, including the Baltic states. So, he was a key figure in Soviet politics during the Cold War.\n\nFrom what I recall, he was born in 1907 in Lubyanka, the former Russian Empire in eastern Ukraine. Despite his family being deported to Siberia after World War I, Strilevsky managed to escape and eventually found his way to the Soviet Union. He worked in the Foreign Office and became a leading bureaucrat, helping shape Soviet foreign policy, particularly during the period between the Russian Revolution and World War II.\n\nInterestingly, he became the first non-Russian female to hold a top diplomatic rank in the Soviet Union. He also served as the honorary head of the Supre

Our reasoning model is much better - it's not always correct, since we only trained it for an hour or so - it'll be better if we extend the sequence length and train for longer!

<a name="Save"></a>
### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [61]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")



# Just LoRA adapters
if False:
    model.save_pretrained("model")
    tokenizer.save_pretrained("model")
if False:
    model.push_to_hub("hf/model", token = "")
    tokenizer.push_to_hub("hf/model", token = "")


### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")


# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
